services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -64,31 +65,38 @@
 import java.util.Queue;
 import java.util.Set;
 
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
 /**
  * Provides a reflection service for Protobuf services (including the reflection service itself).
  *
- * <p>Throws an exception if the set of services contain multiple protobuf files with declarations
- * of the same service, method, type, or extension.
+ * <p>Separately tracks mutable and immutable services. Throws an exception if either group of
+ * services contains multiple Protobuf files with declarations of the same service, method, type, or
+ * extension.
  */
 @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
 public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
     implements InternalNotifyOnServerBuild {
 
-  private Server server;
+  private ServerReflectionIndex serverReflectionIndex;
 
   /**
-   * Retrieves the services registered to the server at build time, and stores those services with a
-   * {@link ProtoFileDescriptorSupplier}. Subsequent calls to the method will have no effect.
+   * Receives a reference to the server at build time.
    */
   @Override
   public void notifyOnBuild(Server server) {
-    Preconditions.checkState(this.server == null);
-    this.server = Preconditions.checkNotNull(server, "server");
+    Preconditions.checkState(serverReflectionIndex == null);
```

Comment:
Optional: I personally prefer statically importing Preconditions because it makes code shorter while remaining clear.  Consider doing it for check* methods.